poly/src/domain/radix2/fft.rs

```diff
@@ -20,6 +20,46 @@ enum FFTOrder {
 }
 
 impl<F: FftField> Radix2EvaluationDomain<F> {
+    /// Degree aware FFT that runs in O(n log d) instead of O(n log n)
+    /// Implementation copied from libiop.
+    pub(crate) fn degree_aware_fft_in_place<T: DomainCoeff<F>>(&self, coeffs: &mut Vec<T>) {
+        if !self.offset.is_one() {
+            Self::distribute_powers(&mut *coeffs, self.offset);
+        }
+        let n = self.size();
+        let log_n = self.log_size_of_group;
+        let num_coeffs = coeffs.len();
+        let log_d = ark_std::log2(num_coeffs);
+        // When the polynomial is of size k*|coset|, for k < 2^i,
+        // the first i iterations of Cooley Tukey are easily predictable.
+        // This is because they will be combining g(w^2) + wh(w^2), but g or h will always refer
+        // to a coefficient that is 0.
+        // Therefore those first i rounds have the effect of copying the evaluations into more locations,
+        // so we handle this in initialization, and reduce the number of loops that are performing arithmetic.
+        // The number of times we copy each initial non-zero element is as below:
+        let duplicity_of_initials = 1 << (log_n - log_d);
```

Comment:
Let's make this is a `checked_sub`, or maybe handle the case where `d >= n` separately.